Provide incremental by-subsystem initialization for the crypto library

athoelke opened the pull request:
See #16 for discussion of this API, which enables partial initialization of the library. This is useful in constrained contexts, for example during early boot, when not all library functionality is required to support the application use case.

Fixes #16

gilles-peskine-arm reviewed (changes requested):
This design was prototyped in https://github.com/Mbed-TLS/mbedtls/pull/6636 and I consider that prototype to be sufficient validation for the API design.

The specification looks good to me except for a few places where some Mbed TLS specific history remains.

athoelke commented on doc/crypto/api/keys/attributes.rst:
```diff
@@ -159,7 +159,7 @@ Managing key attributes
     .. retval:: PSA_ERROR_DATA_INVALID
     .. retval:: PSA_ERROR_BAD_STATE
-        The library requires initializing by a call to `psa_crypto_init()`.
+        The library requires initializing. See :secref:`library-init`.
```
That is a likely explanation - this PR has been here for a bit, I may have rebased this after other PRs were merged without updating the set of changes like this.

I will rebase again, and then update any additional APIs that need this text.

athoelke commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +83,156 @@ Library initialization
+    .. summary::
+        Crypto subsystem identifier for the communication with the server, if this is a client that communicates with a server where the key store is located.
+
+    In a client-server implementation, initializing this subsystem is necessary before any API function other than library initialization, deinitialization and functions accessing local data structures such as key attributes.
```
Thank you for spotting that

athoelke commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +83,156 @@ Library initialization
+    .. summary::
+        Crypto subsystem identifier for accelerator drivers.
+
+    Initializing this subsystem results in initialization of all registered accelerator drivers.
```
I think that is worth a small PR of its own to discuss such things somewhere between the design goals/sample architectures/implementation considerations?

athoelke commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +83,156 @@ Library initialization
+    :definition: /* implementation-defined value */
+
+    .. summary::
+        Crypto subsystem identifier for all available subsystems.
```
I think this is effectively implied by the rule that using this identifier in a call to `psa_crypto_init_subsystem()` is equivalent to a call to `psa_crypto_init()`. But maybe we should spell that out, to avoid an implementation short cut that doesn't consider this kind of time evolution of the implementation?

athoelke commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +83,156 @@ Library initialization
+    It is guaranteed that the following operations do not to require this subsystem:
+
+    *   Hash operations.
+    *   Signature verification operations.
```
I've added key encapsulation and PAKE to the list of operations that require the RANDOM subsystem; and modified the 'some operations using the private or secret key' statement to 'some operations using key pairs' to more readily include export of public keys (where this is calculated on demand).

I can add attribute retrieval, and also export of a symmetric key type or a public key type, to the list of not-needing randomness.

athoelke commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +83,156 @@ Library initialization
+    It is guaranteed that the following operations do not to require this subsystem:
+
+    *   Hash operations.
+    *   Signature verification operations.
```
The scenario that a subsystem might be initialized implicitly as another subsystem depends on it is described in the description of `psa_crypto_init_subsystem()`, using the example of communications in a client/server implementation. We could add this example to that description, or we could add a note that RANDOM may be required by other subsystems, for example to initialize communication with a secure element.

athoelke commented:
Rebased and updated this PR

athoelke commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +83,156 @@ Library initialization
+    :definition: /* implementation-defined value */
+
+    .. summary::
+        Crypto subsystem identifier for access to built-in keys.
```
This might be worth adding to a separate PR, alongside the coverage of accelerators and secure elements

gilles-peskine-arm reviewed (commented):
I've reviewed up to 5360369a447d87caead0d01206840542e427a68f. Ok so far. Still to be addressed:

* [x] https://github.com/ARM-software/psa-api/pull/202#discussion_r1827940872
* [x] https://github.com/ARM-software/psa-api/pull/202#discussion_r1827950757
* [x] https://github.com/ARM-software/psa-api/pull/202#discussion_r1827951963
* [x] https://github.com/ARM-software/psa-api/pull/202#discussion_r1827953963
* [x] https://github.com/ARM-software/psa-api/pull/202#discussion_r1827969110

athoelke commented:
> I've reviewed up to [5360369](https://github.com/ARM-software/psa-api/commit/5360369a447d87caead0d01206840542e427a68f). Ok so far. Still to be addressed:

I think I've covered all except the need for some discussion of accelerators, secure elements, and built-in keys within the document.

I plan to deal with that in a separate PR, unless you would prefer to keep it coupled with this one?

athoelke commented:
Ah, I realise that [this one](https://github.com/ARM-software/psa-api/pull/202#discussion_r1827969110) still needs a change to the PR.

gilles-peskine-arm commented:
Ok for a separate PR. But I don't think considerations around “drivers” and “built-in keys” should leak outside the specification of subsystem initialization for now.

athoelke commented:
> Ok for a separate PR. But I don't think considerations around “drivers” and “built-in keys” should leak outside the specification of subsystem initialization for now.

I think I could reword the text for the subsystem definitions to avoid talking about 'drivers' - which suggests a specific way of a portable implementation providing support for varied hardware. I just need to review the spec for a mention of 'secure element' and 'accelerator', and add/amend if necessary.

athoelke commented:
I've added a couple more commits.

* In one I add some terms to the glossary for accelerators, secure elements and built-int keys - using these in appropriate places.
* I have removed 'registering drivers' from the text in the subsystem initialization
* In the other, I add a recommendation for implementation of 'all subsystem' initialization.

gilles-peskine-arm commented on doc/crypto/api/library/library.rst:
```diff
@@ -52,24 +84,191 @@ Library initialization
+    See also :secref:`library-init`.
 
-    *   If failure to initialize the library does not compromise the security of the function, the implementation must either provide the expected result for the function, or return :code:`PSA_ERROR_BAD_STATE` or other appropriate error.
+.. typedef:: uint32_t psa_crypto_subsystem_t
```
In my [prototype](https://github.com/Mbed-TLS/mbedtls/pull/6636/changes#diff-8fe19c51d861f48010b3609673ee1dd94bb3eef20af920af31b36fc0dcb124c8R220), I also had way to query the initialization state of subsystems. I implemented [`int mbedtls_psa_crypto_is_subsystem_initialized(psa_crypto_subsystem_t)` which returns a boolean result indicating whether the given subsystem mask is fully initialized. Another possibility would be a function that returns a subsystem mask indicating what is currently initialized. Would this be desirable as a standard function?

athoelke replied:
Perhaps that makes sense. An `is_initialized()` boolean test matches the `init()` signature which is very tidy for code that wants to check - on the other hand just calling `init(subset)` on the same set of subsystems should always be quicker than `if (!is_initialized(subset)) init(subset);`.

So if we add an API, it might make more sense to be one that reports the set of initialized subsystems?
